### PR TITLE
PARQUET-384: Add dictionary filtering.

### PR DESCRIPTION
**Pull request by rdblue**
This builds on #286 from @danielcweeks and cleans up some of the interfaces. It introduces `DictionaryPageReadStore` to expose dictionary pages to the filters and cleans up some internal calls by passing `ParquetFileReader`.

When committed, this closes #286.

### CODE REVIEW

File: parquet-hadoop/src/main/java/org/apache/parquet/filter2/dictionarylevel/DictionaryFilter.java [hunk 1]
```diff
@@ -0,0 +1,356 @@
...
+    filterColumn.getColumnPath();
+
+    try {
+      Set<T> dictSet = expandDictionary(meta);
+      if (dictSet == null) {
+        return BLOCK_MIGHT_MATCH;
+      }
+
+      for(T entry : dictSet) {
+        if(value.compareTo(entry) > 0) {
```
Comment (danielcweeks) [inline parquet-hadoop/src/main/java/org/apache/parquet/filter2/dictionarylevel/DictionaryFilter.java]: Shouldn't this be `>=`?

File: parquet-hadoop/src/main/java/org/apache/parquet/filter2/dictionarylevel/DictionaryFilter.java [hunk 2]
```diff
@@ -0,0 +1,356 @@
...
+    filterColumn.getColumnPath();
+
+    try {
+      Set<T> dictSet = expandDictionary(meta);
+      if (dictSet == null) {
+        return BLOCK_MIGHT_MATCH;
+      }
+
+      for(T entry : dictSet) {
+        if(value.compareTo(entry) >= 0) {
```
Comment (danielcweeks) [inline parquet-hadoop/src/main/java/org/apache/parquet/filter2/dictionarylevel/DictionaryFilter.java]: Same here `>`?

File: parquet-hadoop/src/main/java/org/apache/parquet/filter2/dictionarylevel/DictionaryFilter.java [hunk 3]
```diff
@@ -0,0 +1,356 @@
...
+    filterColumn.getColumnPath();
+
+    try {
+      Set<T> dictSet = expandDictionary(meta);
+      if (dictSet == null) {
+        return BLOCK_MIGHT_MATCH;
+      }
+
+      for(T entry : dictSet) {
+        if(value.compareTo(entry) < 0) {
```
Comment (danielcweeks) [inline parquet-hadoop/src/main/java/org/apache/parquet/filter2/dictionarylevel/DictionaryFilter.java]: `<=`

File: parquet-hadoop/src/main/java/org/apache/parquet/filter2/dictionarylevel/DictionaryFilter.java [hunk 4]
```diff
@@ -0,0 +1,356 @@
...
+    filterColumn.getColumnPath();
+
+    try {
+      Set<T> dictSet = expandDictionary(meta);
+      if (dictSet == null) {
+        return BLOCK_MIGHT_MATCH;
+      }
+
+      for(T entry : dictSet) {
+        if(value.compareTo(entry) <= 0) {
```
Comment (danielcweeks) [inline parquet-hadoop/src/main/java/org/apache/parquet/filter2/dictionarylevel/DictionaryFilter.java]: `<`

File: parquet-hadoop/src/main/java/org/apache/parquet/filter2/dictionarylevel/DictionaryFilter.java (hunk 1 above)
Comment (rdblue) [inline parquet-hadoop/src/main/java/org/apache/parquet/filter2/dictionarylevel/DictionaryFilter.java]: No, I think this is correct. The logic changes to > because the order is reversed, not because the logic is negated.

If `V` is the bound and we find any value, `x`, in the dictionary such that `x < V`, then there may be a matching row. But we call `V.compareTo(x)` and the order is reversed. We could call `x.compareTo(V)`, but it seems like calling the method on `V` is more likely to result in something the JVM can optimize.

The tests also validate that this behavior is correct by testing the boundary conditions with the smallest and largest values.

File: parquet-hadoop/src/main/java/org/apache/parquet/hadoop/InternalParquetRecordReader.java [hunk 5]
```diff
@@ -163,46 +156,26 @@ public float getProgress() throws IOException, InterruptedException {
...
-    reader = new ParquetFileReader(configuration, parquetFileMetadata, file, blocks, columns);
-    for (BlockMetaData block : blocks) {
-      total += block.getRowCount();
-    }
+    this.total = reader.getRecordCount();
     this.unmaterializableRecordCounter = new UnmaterializableRecordCounter(configuration, total);
     LOG.info("RecordReader initialized will read a total of " + total + " records.");
   }
 
-  private boolean contains(GroupType group, String[] path, int index) {
```
Comment (rdblue) [inline parquet-hadoop/src/main/java/org/apache/parquet/hadoop/InternalParquetRecordReader.java]: This method wasn't used.

**Comment by rdblue** [comment]
I just added caching and a metadata check to the dictionary reader so that dictionary pages are only read once even if used in multiple predicates (which each call `readDictionary`).

File: parquet-hadoop/src/main/java/org/apache/parquet/filter2/dictionarylevel/DictionaryFilter.java (hunk 1 above)
Comment (danielcweeks) [inline parquet-hadoop/src/main/java/org/apache/parquet/filter2/dictionarylevel/DictionaryFilter.java]: Yeah, you're right.

File: parquet-hadoop/src/main/java/org/apache/parquet/hadoop/ParquetFileReader.java [hunk 6]
```diff
@@ -547,19 +675,113 @@ public PageReadStore readNextRowGroup() throws IOException {
...
       for (Chunk chunk : chunks) {
-        columnChunkPageReadStore.addColumn(chunk.descriptor.col, chunk.readAllPages());
+        currentRowGroup.addColumn(chunk.descriptor.col, chunk.readAllPages());
       }
     }
+
+    // avoid re-reading bytes the dictionary reader is used after this call
+    if (nextDictionaryReader != null) {
+      nextDictionaryReader.setRowGroup(currentRowGroup);
+    }
```
Comment (danielcweeks) [inline parquet-hadoop/src/main/java/org/apache/parquet/hadoop/ParquetFileReader.java]: `advanceToNextBlock` resets nextDictionaryReader.  Does that mean the dictionary reader is not available after the row group read?

Comment (rdblue) [inline parquet-hadoop/src/main/java/org/apache/parquet/hadoop/ParquetFileReader.java]: `advanceToNextBlock` resets the next reader because the next block becomes the current block. This happens inside skip or read. The expected use is to get the dictionary reader for the next block, determine whether to skip or read it, do that operation, and then get the reader for the next block. Skipping or reading should cause the file reader to prepare to return the dictionary reader for the next, which is what is happening here.

**Comment by rdblue** [comment]
#332 is a follow-up to this and adds EncodingStats so to avoid the brittle logic in `hasNonDictionaryEncodedPages`.

**Comment by danielcweeks** [comment]
+1  The ParquetFileReader has some general issues with how it should be used, but I think this works really well for the dictionary filter and cleaning up some of the block filtering.
